addons/addon-base-raas/packages/base-raas-services/lib/data-source/access-strategy/roles-only/application-role-service.js

```diff
@@ -0,0 +1,263 @@
+   * is not found.
+   *
+   * @param requestContext The standard requestContext
+   * @param accountId The data source account id that the application role belongs to
+   * @param bucket The name of the bucket
+   * @param arn The application role arn
+   * @param fields An array of the attribute names to return, default to all the attributes
+   * of the application role entity.
+   */
+  async mustFind(requestContext, { accountId, bucket, arn, fields = [] }) {
+    const result = await this.find(requestContext, { accountId, bucket, arn, fields });
+    if (!result) throw this.boom.notFound(`Application role with arn "${arn}" does not exist`, true);
+    return result;
+  }
+
+  /**
+   * Call this method to allocate an application role entity for the given study. This method is smart
+   * enough to reuse an existing app role entity if there is enough space for the study to fit in.
+   *
+   * Note: this method does NOT actually create an IAM role resource in an AWS account.
+   *
+   * @param requestContext The standard requestContext
+   * @param accountEntity The data source account entity
+   * @param bucketEntity The data source bucket entity
+   * @param studyEntity The study entity
+   */
+  async allocateRole(requestContext, _accountEntity = {}, bucketEntity = {}, studyEntity = {}) {
+    // Allocating an application role is only applicable for bucket with access = 'roles'
+    if (studyEntity.bucketAccess !== 'roles') return undefined;
+
+    // Perform default condition checks to make sure the user is active
+    await this.assertAuthorized(
+      requestContext,
+      { action: 'allocate', conditions: [allowIfActive, allowIfAdmin] },
+      studyEntity,
+    );
+    const { accountId, id: studyId, bucket } = studyEntity;
+
+    // The logic
+    // - Get all the application roles entities by the account and the bucket
+    // - Do we have the study already part of an existing application role?
+    //   - if yes, then return the application role
+    // - For each application roles, do the following:
+    //   - add study to the role entity
+    //   - check if size has reached the max
+    //    - if not then we found a place for the study and we update the application role
+    //      entity and return it
+    // - If none is found, create a new application role, add study to it, store it
+    //   in the database  and return the application role entity
+
+    const appRoleEntities = await this.list(requestContext, accountId, { bucket });
+    let appRoleEntity = _.find(appRoleEntities, entity => _.has(entity, ['studies', studyId]));
+
+    if (!_.isUndefined(appRoleEntity)) return appRoleEntity;
+
+    appRoleEntity = undefined;
+    // eslint-disable-next-line consistent-return
+    _.forEach(appRoleEntities, entity => {
+      if (!maxReached(addStudy(entity, studyEntity))) {
+        appRoleEntity = entity;
+        return false; // To stop the loop since we found a role
+      }
+    });
```

Comment:
Sure, I can do that. Let me first see if I no longer need any more logic inside line 145, since this is still WIP.

Reply:
So can we resolve this for now and it will be addressed in the follow up PR.

Reply:
When I looked at the logic again, I realized that it might be better to leave it as a foreach loop rather than find. The main difference is that when I think of find, I think of a method that does not mutate anything.  While in my case, I need to add the study to the role entity before I can determine if the max size has been reached. I am thinking it might be better to leave it as a foreach loop.